src/mfs.jl

```diff
@@ -0,0 +1,230 @@
+`mapper` one of available in Attractors.jl [`AttractorMapper`](@ref) constructed with respect to dynamical system
+`X` initial point to be tested
+`best_shock` best shock found by `random_pertubation`
+`best_dist` norm of the best shock found by `random_pertubation`
+`dim` dimension of the system
+`total_iterations` number of random pertubations to try, default = 10000
+`delta` factor by which the radius of the hypersphere is decreased, default = 1000.0, the higher the value
+
+
+"""
+function mfs_brute_force(mapper::AttractorMapper, X, 
+                        best_shock, best_dist, dim, id_x,  
+                        total_iterations=10000, delta = 1000.0)
+
+    temp_dist = best_dist-best_dist/delta
+    perturbation = zeros(dim)
+
+    for _ in 1:total_iterations
+        generator, _ = statespace_sampler(GLOBAL_RNG;
+                                         radius = 1.0, spheredims = dim, 
+                                         center = zeros(dim) ) 
+        perturbation = generator() * temp_dist
+        new_shock = perturbation + X
+        
+        if !(id_x == mapper(new_shock))
+            best_dist = norm(perturbation)
+            best_shock = perturbation
+            temp_dist = best_dist-best_dist/delta
+        end
+    end
+
+    return best_shock, best_dist
+end
+
+
+
+
+function mfs_objective(perturbation, X, idX, mapper::AttractorMapper, penalty=1000.0) 
+
+    dist = norm(perturbation)
+    if dist == 0
+        return penalty
+    end
+    if mapper(X + perturbation) == idX
+        return dist + penalty
+    else
+        return dist
+    end
+end
+
+
+"""
+    minimal_fatal_shock(mapper::AttractorMapper, X, search_area, algorithm::Union{MFSBruteForce, MFSBlackBoxOptim})  
+
+Runs the minimal fatal shock algorithm on the initial point X. Two algorithms are available: MFSBruteForce and MFSBlackBoxOptim.
+MFSBruteForce is an algorithm based on randomised search with two consequent steps: random initialization and sphere radius reduction.
+MFSBlackBoxOptim is an algorithm based derivative free optimization. It uses BlackBoxOptim package to find the best shock.
+
+`mapper::AttractorMapper` one of available in Attractors.jl [`AttractorMapper`](@ref) constructed with respect to dynamical system.
+`X` vector containing coordinates of initial point to be tested.
+`search_area` array of tuples, by specifying single value e.g. [(-1.5, 1.5)] you specify this value for each of the variables, or you may specify separetely 
+search area for each of your states by, for example in 2 dimensions, [(-1.5, 1.5), (-1, 1)].
+`algorithm` one of the two algorithms available: `MFSBruteForce` or `MFSBlackBoxOptim`.
+
+## `MFSBruteForce`
+`initial_iterations` number of random pertubations to try in the first step of the algorithm, default = 10000.
+`sphere_iterations` number of steps while initializing random points on hypersphere and decreasing its radius, default = 10000.
+`dimension` dimension of the system, default = 2.
```

Comment:
Why would need a dimension, and how can the dimension possibly have a default value? Don't ask the user to provide you with information you already have. You have the dimension information from the dynamical system. Additionally, information that _cannot_ have a default value cannot be a keyword argument. Here dimension cannot have a default value. There is not way to pick a default. Thankfully, we don't have to, as we already have access to the dimension information.

Reply:
`dimension(ds::DynamicalSystem)` is the function you need.

Reply:
`dimension` now is computed on the entry to main function `minimal_fatal_shock` but i still leave it in auxiliary  function as argument to avoid double computations